In [107]:
import numpy as np
import pandas as pd

# 1 Data analysis

1. Import the data set. Make sure the column names are imported as intended.

In [108]:
path = "C:/Users/User/Documents/GitHub/CausalAI-Course-1/"
data = pd.read_csv(path + 'data/wage2015_subsample_inference.csv')
list(data.columns)


['rownames',
 'wage',
 'lwage',
 'sex',
 'shs',
 'hsg',
 'scl',
 'clg',
 'ad',
 'mw',
 'so',
 'we',
 'ne',
 'exp1',
 'exp2',
 'exp3',
 'exp4',
 'occ',
 'occ2',
 'ind',
 'ind2']

In [118]:
data = data.drop('rownames',axis=1)
data

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we   ne  \
0      9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
1     48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
2     11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
3     13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
4     28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
...         ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
5145  14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
5146  23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0   
5147  38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
5148  32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
5149  17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   

      exp1  exp2    exp3     exp4     occ  occ2     ind  ind2  
0      7.0  0.49   0.343   0.2401  3600.0    11  8370.0    18  
1     31.0  9.61  29.791  92.3521  3050.0    10  5070.0     9  
2     18.0  3.24   5.832  10.4976  6260.0    19   770.0     4  
3     25.0  6.25  15.625  39.0625   420.0     1  6990.0    12  
4     22.0  4.84  10.648  23.4256  2015.0     6  9470.0    22  
...    ...   ...     ...      ...     ...   ...     ...   ...  
5145   9.0  0.81   0.729   0.6561  4700.0    16  4970.0     9  
5146  12.0  1.44   1.728   2.0736  4110.0    13  8680.0    20  
5147  11.0  1.21   1.331   1.4641  1550.0     4  3680.0     6  
5148  10.0  1.00   1.000   1.0000  2920.0     9  6570.0    11  
5149  14.0  1.96   2.744   3.8416  1610.0     5  7460.0    14  

[5150 rows x 20 columns]

2. Are there missing values? Display the number of missings (if any) of each variable in the data set.

In [119]:
data.isna().sum()

wage     0
lwage    0
sex      0
shs      0
hsg      0
scl      0
clg      0
ad       0
mw       0
so       0
we       0
ne       0
exp1     0
exp2     0
exp3     0
exp4     0
occ      0
occ2     0
ind      0
ind2     0
dtype: int64

3. Report descriptive statistics of the variables (mean, standard deviation, percetiles, etc.). Interpret your results.

In [120]:
data.describe()

wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.216117     0.227767    13.760583     3.018925     8.235867   
std       0.411635     0.419432    10.609465     4.000904    14.488962   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     5.000000     0.250000     0.125000   
50%       0.000000     0.000000    10.000000     1.000000     1.000000   
75%       0.000000     0.000000    21.000000     4.410000     9.261000   
max       1.000000     1.000000    47.000000    22.090000   103.823000   

              exp4            occ         occ2            ind         ind2  
count  5150.000000    5150.000000  5150.000000    5150.000000  5150.000000  
mean     25.118038    5310.737476    11.670874    6629.154951    13.316893  
std      53.530225   11874.356080     6.966684    5333.443992     5.701019  
min       0.000000      10.000000     1.000000     370.000000     2.000000  
25%       0.062500    1740.000000     5.000000    4880.000000     9.000000  
50%       1.000000    4040.000000    13.000000    7370.000000    14.000000  
75%      19.448100    5610.000000    17.000000    8190.000000    18.000000  
max     487.968100  100000.000000    22.000000  100000.000000    22.000000

In [124]:
# more percentiles included

prctile = np.arange(0,1,0.05)
data.describe(percentiles=prctile)


wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
0%        3.021978     1.105912     0.000000     0.000000     0.000000   
5%        8.173077     2.100845     0.000000     0.000000     0.000000   
10%       9.615385     2.263364     0.000000     0.000000     0.000000   
15%      11.207532     2.416586     0.000000     0.000000     0.000000   
20%      12.019231     2.486508     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
30%      14.423077     2.668829     0.000000     0.000000     0.000000   
35%      15.384615     2.733368     0.000000     0.000000     0.000000   
40%      16.826923     2.822980     0.000000     0.000000     0.000000   
45%      17.713942     2.874352     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
55%      20.192308     3.005302     0.000000     0.000000     0.000000   
60%      21.634615     3.074295     1.000000     0.000000     0.000000   
65%      24.038462     3.179655     1.000000     0.000000     0.000000   
70%      25.641026     3.244194     1.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
80%      30.364372     3.413270     1.000000     0.000000     1.000000   
85%      34.188034     3.531876     1.000000     0.000000     1.000000   
90%      39.615385     3.679218     1.000000     0.000000     1.000000   
95%      50.000000     3.912023     1.000000     0.000000     1.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
0%        0.000000     0.000000     0.000000     0.000000     0.000000   
5%        0.000000     0.000000     0.000000     0.000000     0.000000   
10%       0.000000     0.000000     0.000000     0.000000     0.000000   
15%       0.000000     0.000000     0.000000     0.000000     0.000000   
20%       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
30%       0.000000     0.000000     0.000000     0.000000     0.000000   
35%       0.000000     0.000000     0.000000     0.000000     0.000000   
40%       0.000000     0.000000     0.000000     0.000000     0.000000   
45%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
55%       0.000000     0.000000     0.000000     0.000000     0.000000   
60%       0.000000     0.000000     0.000000     0.000000     0.000000   
65%       0.000000     0.000000     0.000000     0.000000     0.000000   
70%       0.000000     1.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
80%       1.000000     1.000000     0.000000     1.000000     1.000000   
85%       1.000000     1.000000     0.000000     1.000000     1.000000   
90%       1.000000     1.000000     1.000000     1.000000     1.000000   
95%       1.000000     1.000000     1.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.2161

As most of the variables are dummy variables, only wage is interprated based in these descriptive statistics. The average wage in the sample is 23.410 dollars per hour. The 25% richest earn 27.78 dollars per hour or more, while the 25% poorest earn 13.45 or less. The median earning per hour is 19.23 dollars, which is lower than the average wage and shows the asymmetry in the wage distribution of the sample. 

4. How many women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [125]:
data_women_clg_75 = data[(data['sex'] == 1) &
                         ((data['clg'] == 1)|(data['ad'] == 1)) &
                         (data['wage']>data['wage'].describe()['75%'])]
data_women_clg_75

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we   ne  \
4      28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
30     42.307692  3.744969  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
39     41.208791  3.718652  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
54    100.000000  4.605170  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
91     33.653846  3.516127  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
...          ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
5085   48.076923  3.872802  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
5093   30.145530  3.406037  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
5105   29.914530  3.398344  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
5131   48.076923  3.872802  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
5140   45.546559  3.818735  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   

      exp1   exp2    exp3      exp4     occ  occ2     ind  ind2  
4     22.0   4.84  10.648   23.4256  2015.0     6  9470.0    22  
30     0.0   0.00   0.000    0.0000  4700.0    16  4795.0     9  
39    11.0   1.21   1.331    1.4641  4700.0    16  5790.0     9  
54    27.0   7.29  19.683   53.1441   850.0     2  6970.0    12  
91     4.0   0.16   0.064    0.0256   710.0     2  7390.0    14  
...    ...    ...     ...       ...     ...   ...     ...   ...  
5085   0.0   0.00   0.000    0.0000  3050.0    10  8090.0    18  
5093  15.0   2.25   3.375    5.0625   120.0     1  6870.0    12  
5105  39.0  15.21  59.319  231.3441  4700.0    16  5380.0     9  
5131  25.0   6.25  15.625   39.0625  3255.0    10  8170.0    18  
5140   5.0   0.25   0.125    0.0625  3255.0    10  8190.0    18  

[412 rows x 20 columns]

In [126]:
print(f'{data_women_clg_75.shape[0]} women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample')

412 women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample


5. How many men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [175]:
data_men_hsg_75 = data[(data['sex'] == 1) &
                       ((data['shs'] == 1)|(data['hsg'] == 1)) &
                       (data['wage']>data['wage'].describe()['75%'])]
data_men_hsg_75

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we   ne  \
15    27.884615  3.328075  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
46    28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
75    28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
136   28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
140   29.714286  3.391628  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
...         ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
5074  33.653846  3.516127  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
5086  40.865385  3.710283  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
5089  31.250000  3.442019  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
5094  33.653846  3.516127  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
5148  32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   

      exp1   exp2    exp3      exp4     occ  occ2     ind  ind2  
15    16.0   2.56   4.096    6.5536  6355.0    19   770.0     4  
46    41.0  16.81  68.921  282.5761  9620.0    22  4970.0     9  
75    33.0  10.89  35.937  118.5921  8220.0    21  3680.0     6  
136   40.0  16.00  64.000  256.0000  4700.0    16  4870.0     9  
140    8.0   0.64   0.512    0.4096  4040.0    13  8680.0    20  
...    ...    ...     ...       ...     ...   ...     ...   ...  
5074  30.0   9.00  27.000   81.0000  6320.0    19   770.0     4  
5086  33.0  10.89  35.937  118.5921  7140.0    20  6070.0    10  
5089  34.0  11.56  39.304  133.6336  5400.0    17  9370.0    22  
5094  28.0   7.84  21.952   61.4656  9030.0    22  6070.0    10  
5148  10.0   1.00   1.000    1.0000  2920.0     9  6570.0    11  

[115 rows x 20 columns]

In [176]:
print(f'{data_men_hsg_75.shape[0]} men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample')

115 men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample


6. Create two dataframes. One containing only the log(wage) and the other containig every variable of the data set but the wage related variables.

In [129]:
log_wage = pd.DataFrame(data['lwage'])
log_wage


lwage
0     2.263364
1     3.872802
2     2.403126
3     2.634928
4     3.361977
...        ...
5145  2.692546
5146  3.138833
5147  3.649659
5148  3.495508
5149  2.851151

[5150 rows x 1 columns]

In [174]:
vars = [x for x in data.columns if x != 'wage' and x !='lwage']
features = pd.DataFrame(data[vars])
features

sex  shs  hsg  scl  clg   ad   mw   so   we   ne  exp1  exp2    exp3  \
0     1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   7.0  0.49   0.343   
1     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  31.0  9.61  29.791   
2     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  18.0  3.24   5.832   
3     1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  25.0  6.25  15.625   
4     1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  22.0  4.84  10.648   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   ...     ...   
5145  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   9.0  0.81   0.729   
5146  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  12.0  1.44   1.728   
5147  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  11.0  1.21   1.331   
5148  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  10.0  1.00   1.000   
5149  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  14.0  1.96   2.744   

         exp4     occ  occ2     ind  ind2  
0      0.2401  3600.0    11  8370.0    18  
1     92.3521  3050.0    10  5070.0     9  
2     10.4976  6260.0    19   770.0     4  
3     39.0625   420.0     1  6990.0    12  
4     23.4256  2015.0     6  9470.0    22  
...       ...     ...   ...     ...   ...  
5145   0.6561  4700.0    16  4970.0     9  
5146   2.0736  4110.0    13  8680.0    20  
5147   1.4641  1550.0     4  3680.0     6  
5148   1.0000  2920.0     9  6570.0    11  
5149   3.8416  1610.0     5  7460.0    14  

[5150 rows x 18 columns]